SSA/WellTypedFramework.lean

```diff
@@ -1,91 +1,128 @@
+  | some : BaseType → TypeData BaseType
+  | unit : TypeData BaseType
+  | pair : TypeData BaseType → TypeData BaseType → TypeData BaseType
+  | triple : TypeData BaseType → TypeData BaseType → TypeData BaseType → TypeData BaseType
+  | any : TypeData BaseType
+  | unused : TypeData BaseType -- Or bound
+
+@[coe]
+def UserType.toTypeData [DecidableEq BaseType] [BaseTypeClass BaseType] : UserType BaseType → TypeData BaseType
+  | UserType.base t => TypeData.some t
+  | UserType.unit => TypeData.unit
+  | UserType.pair k₁ k₂ => TypeData.pair (UserType.toTypeData k₁) (UserType.toTypeData k₂)
+  | UserType.triple k₁ k₂ k₃ => TypeData.triple (UserType.toTypeData k₁) (UserType.toTypeData k₂)
+    (UserType.toTypeData k₃)
+
+variable {BaseType : Type} [instDecidableEqBaseType : DecidableEq BaseType] [instBaseTypeClass : BaseTypeClass BaseType]
+instance : Coe (UserType BaseType) (TypeData BaseType) := ⟨UserType.toTypeData⟩
+
+class TypedUserSemantics (Op : Type) (BaseType : Type) [DecidableEq BaseType] [BaseTypeClass BaseType] where
+  argKind : Op → UserType BaseType
+  rgnDom : Op → UserType BaseType
+  rgnCod : Op → UserType BaseType
+  outKind : Op → UserType BaseType
+  eval : ∀ (o : Op), (argKind o).toType → ((rgnDom o).toType →
+    (rgnCod o).toType) → (outKind o).toType
 
 @[simp]
-def TypeData.toType (TS : TypeSemantics) : TypeData → Type
-  | TypeData.some k => TS k
+def TypeData.toType : TypeData BaseType → Type
+  | TypeData.some t => BaseTypeClass.toType t
   | TypeData.unit => Unit
-  | TypeData.pair t₁ t₂ => t₁.toType TS × t₂.toType TS
-  | TypeData.triple t₁ t₂ t₃ => t₁.toType TS × t₂.toType TS × t₃.toType TS
-  | TypeData.any => Σ (k : Kind), k.toType TS
+  | TypeData.pair t₁ t₂ => (TypeData.toType  t₁) × (TypeData.toType t₂)
+  | TypeData.triple t₁ t₂ t₃ => (TypeData.toType t₁) × (TypeData.toType t₂) × (TypeData.toType t₃)
+  | TypeData.any => Σ (k : UserType BaseType), @UserType.toType BaseType instDecidableEqBaseType instBaseTypeClass k
   | TypeData.unused => Unit
 
+theorem TypeDataToTypeEqUserTypeToType (a : UserType BaseType) : a.toType = TypeData.toType a.toTypeData :=
+  by
+    induction a <;> (try rfl)
+    case pair k₁ k₂ ih₁ ih₂ => simp [TypeData.toType, UserType.toType, ih₁, ih₂]
+    case triple k₁ k₂ k₃ ih₁ ih₂ ih₃ => simp [TypeData.toType, UserType.toType, ih₁, ih₂, ih₃]
+
+
 @[simp]
-def SSAIndex.ieval (Op : Type) (TS : TypeSemantics)
-    [TypedUserSemantics Op TS] : SSAIndex → Type
-  | .TERMINATOR => TypeData
-  | .EXPR => TypeData
-  | .REGION => TypeData × TypeData
+def SSAIndex.ieval (Op : Type)
+    [TypedUserSemantics Op BaseType] : SSAIndex → Type
+  | .TERMINATOR => TypeData BaseType
+  | .EXPR => TypeData BaseType
+  | .REGION => TypeData BaseType × TypeData BaseType
   | .STMT => List Var
 
-abbrev K : Type → Type := StateT ((Var → TypeData) × (Var → (TypeData × TypeData))) Option
-
+abbrev K : Type → Type := StateT ((Var → TypeData BaseType) × (Var → (TypeData  BaseType × TypeData BaseType))) Option
```

Comment:
I suggest parameterizing K by `BaseType`. That should allow Lean to figure out what we mean, as then `BaseType` becomes an explicit argument (instead of an implicit as it is right now)

Reply:
no octopus emoji in github :(